### PR TITLE
Improve speed of Golang-related Makefile targets when running on macOS (and don't break Linux hosts)

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

Same as #3527 + a fix for Linux hosts (see https://github.com/grafana/mimir/pull/3527#issuecomment-1330810445 and #3569)

I don't have permission to push an updated build image - could a maintainer please follow the instructions in https://github.com/grafana/mimir/blob/1806872c575eeaaaab132cce55cd157360cfdaca/docs/internal/how-to-update-the-build-image.md to build, push and update the image reference everywhere?

#### Which issue(s) this PR fixes or relates to

n/a

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by charleskorn** [comment]
That's expected @jhesketh - I've had to make a change to `build.sh`, which requires a new build image to be built and pushed (and then references to the build image to be updated to the new one). Unfortunately I don't have access to push the image myself.

If you'd like to test this locally before the image is updated, these rough steps should work:
1. Build the build image: `docker build --build-arg=goproxyValue=https://proxy.golang.org,direct --build-arg=revision=does-not-matter -t grafana/mimir-build-image:local-testing mimir-build-image`
2. Disable pulling the build image in `Makefile`: comment out the `docker pull` step in the `fetch-docker-image` target
3. Run `LATEST_BUILD_IMAGE_TAG=local-testing make test`

(there may be a nicer way to test this locally)

**Comment by jhesketh** [comment]
Ah, of course. Obviously I did not read this through properly. Sorry!

Built the image and confirmed it works for me :+1:

**Comment by pstibrany** [comment]
I'm trying to build and publish build image, but I'm getting:
```
#10 2.237 E: The repository 'http://deb.debian.org/debian bullseye-updates InRelease' is not signed.
#10 ERROR: process "/bin/sh -c apt-get update && apt-get install -y curl python3-requests python3-yaml file jq zip unzip protobuf-compiler libprotobuf-dev shellcheck libpcap-dev skopeo && \trm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*" did not complete successfully: exit code: 100
------
 > [stage-2  2/13] RUN apt-get update && apt-get install -y curl python3-requests python3-yaml file jq zip unzip protobuf-compiler libprotobuf-dev shellcheck libpcap-dev skopeo && 	rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*:
#10 1.025   At least one invalid signature was encountered.
#10 1.563 Err:2 http://deb.debian.org/debian-security bullseye-security InRelease
#10 1.563   At least one invalid signature was encountered.

#10 2.237 W: GPG error: http://deb.debian.org/debian bullseye InRelease: At least one invalid signature was encountered.
#10 2.237 E: The repository 'http://deb.debian.org/debian bullseye InRelease' is not signed.
#10 2.237 W: GPG error: http://deb.debian.org/debian-security bullseye-security InRelease: At least one invalid signature was encountered.
#10 2.237 E: The repository 'http://deb.debian.org/debian-security bullseye-security InRelease' is not signed.
#10 2.237 W: GPG error: http://deb.debian.org/debian bullseye-updates InRelease: At least one invalid signature was encountered.
#10 2.237 E: The repository 'http://deb.debian.org/debian bullseye-updates InRelease' is not signed.
```

Seems unrelated to this change, but not sure what to do about it.

**Comment by pstibrany** [comment]
Googling around, based on https://github.com/debuerreotype/docker-debian-artifacts/issues/122, it could be problem with my Docker version or free disk space in docker. Will try again.

**Comment by pstibrany** [comment]
Updated build image is now published as `docker.io/grafana/mimir-build-image:pr3584-1806872c5`. Please update the PR with references to this image.

**Comment by charleskorn** [comment]
Thanks @pstibrany. I've updated the image reference everywhere, this should be good to go now.

**Review by jhesketh — APPROVED** [review]
lgtm (and verified new image works)

**Review by pstibrany — APPROVED** [review]
Thank you!

**Comment by pstibrany** [comment]
@colega do you want to give this a try before merging?

File: mimir-build-image/build.sh
Comment (pracucci) [inline mimir-build-image/build.sh:18]: [unrelated] This made me realise we could change user name to "mimir" in a follow up PR :)

File: Makefile
Comment (pracucci) [inline Makefile:596]: Isn't this still required if you run `make BUILD_IN_CONTAINER=false packages`?

File: Makefile
Comment (pracucci) [inline Makefile:464]: I think `.cache` is still required for whenever you run `make BUILD_IN_CONTAINER=false ...`

**Comment by pracucci** [comment]
@pstibrany @charleskorn  I've left a couple of comments to ensure we're not breaking `make BUILD_IN_CONTAINER=false`. Could you take a look, please?

File: Makefile
Comment (charleskorn) [reply Makefile:596]: This _seems_ to work fine for me - what needs these directories?

File: Makefile
Comment (charleskorn) [reply Makefile:464]: I can't find any references to `.cache` anywhere (except to exclude it from `find` invocations), and the handful of targets I've just tried with `make BUILD_IN_CONTAINER=false ...` _seem_ to work fine. What needs the `.cache` directory?

**Comment by colega** [comment]
I can confirm this works on my Linux machine. Thank you.

**Comment by charleskorn** [comment]
@pracucci any chance you could take a look at my responses above? Would be good to get this merged if you're OK with it.

File: Makefile
Comment (pracucci) [reply Makefile:596]: These are used by the go build system, but I actually don't think we need to create them (they will be created by go if required).

**Review by pracucci — APPROVED** [review]
LGTM (modulo a nit)

**Comment by pstibrany** [comment]
I will fix remaining nits and merge this, so that I can fix jsonnet in build image. (As found https://github.com/grafana/mimir/pull/3699 and confirmed in https://github.com/grafana/mimir/pull/3702)

File: Makefile
Comment (pracucci) [reply Makefile]: By keep deleting them I mean revert this change and keep:

```
rm -rf -- $(UPTODATE_FILES) $(EXES) .cache dist
```

Comment (charleskorn) [reply Makefile]: > `.cache` is used by the go build system.

This is no longer true (unless I'm missing something). 

Before this PR, if we were building in a container, `.cache` was mounted into the container at `/go/cache`, which is the value of `GOCACHE` (set in the build image Dockerfile [here](https://github.com/grafana/mimir/blob/8f81601f6def539012a58f4ae9fa24d299d3be6c/mimir-build-image/Dockerfile#L59)). If we were building with `BUILD_IN_CONTAINER=false`, the local machine's value of `GOCACHE` would be used (ie. the `.cache` directory was not used).

After this PR, the `BUILD_IN_CONTAINER=false` behaviour remains the same. However, when building in a container, the `mimir-go-cache` Docker volume will be mounted at `/go/cache` and `.cache` will not be used.

Nonetheless, there's still value in cleaning up `.cache` if it was left around from previous builds, so I think it's a good idea to keep `.cache` in the list of directories to remove like you suggest.

**Pull request merged**
